### ISSUE TITLE
Program Mode options

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#6216]
Hi all,

there are a couple of places where `Program` introduces special behavior. I wonder what should be done about these places, in particular due that in a functional design of the core, flags would have to be propagated and that will be cumbersome.

I guess maybe making the program behavior the default would make sense, anyways I'm opening this issue for discussion.

In particular the two places that seem very low in the hierarchy are:

- evarutil.ml:
```ocaml
    if Flags.is_program_mode () then next_name_away na avoid
    else
```

The next part is in `coercion.ml`:
```ocaml
let inh_coerce_to_base ?loc env evd j =
  if Flags.is_program_mode () then
```

Next parts are already in `cases` and pretyping.

**Comment by ppedrot** [comment]
The `Evarutil` use could be removed if we start using open recursion for the pretyper. Essentially, it's just that there are many ways to generate hypothesis names for evars and that Program has a specific, incompatible behaviour w.r.t. vanilla naming.

**Comment by ejgallego** [comment]
My point is more of "do we want to foment such heterogeneous behavior"? Is it justified having several naming generation strategies? It seems to me like that is a recipe for problems.

**Comment by herbelin** [comment]
About "several naming generation strategies", my conclusion at some time was that we should eventually provide a way to support configurable naming strategies at some time. There are many places where we can do better and to be able to do better while preserving the existing, such configurability would help.

**Comment by ejgallego** [comment]
@herbelin a problem with that is that you will need to parameterize _really a lot_ of functions on the naming strategy then. Basically the whole system above `engine` would need to carry a parameter "naming strategy being used".

So IMVHO we should consider the tradeoff on this. Of course, a naming strategy that lives only at the tactic level would be much more reasonable I think.

**Comment by ejgallego** [comment]
Fixed by #9410
